README.md
# Starship Showdown Game
Welcome to Starship Showdown, a thrilling racing game that puts you in the pilot's seat of a futuristic spaceship. Get ready for high-speed action, precision navigation, and intense competition as you navigate your ship through challenging checkpoints to complete laps and emerge as the ultimate space racer.

## Getting Started
1. Go to the [Releases](https://github.com/MBenincasa/starship-showdown/releases) section of this repository.
2. Download the latest release.
3. Extract the contents of the downloaded zip file.
4. Run the `starship-showdown.exe` file.

## Installation
1. Clone the repository or download the ZIP file.
2. Make sure you have [LÖVE](https://love2d.org/) (a 2D game framework) installed on your computer.
3. Run the game using the LÖVE executable: `love path/to/starship-showdown`

## Features
- **Fast-Paced Gameplay:** Experience heart-pounding races as you navigate your spaceship through a series of checkpoints to complete laps.
- **Precision Controls:** Master the controls of your spaceship with responsive handling, using arrow keys to steer, accelerate, brake, and rotate.
- **Strategic Decision-Making:** Choose the best racing lines and execute precise maneuvers to outwit your opponents and secure victory.
- **Dynamic Environments:** Race across a variety of dynamic and visually stunning space environments, each with its own challenges and obstacles.
- **Competitive Racing:** Test your racing skills against AI opponents, aiming to set the fastest lap times and claim the winner's podium.
- **Interactive Settings Menu:** Access an intuitive settings menu to toggle background music, pause the game, and more.
- **AI Opponent:** Compete against advanced AI opponent that provides challenging and engaging races.

## Controls
- Use the arrow keys to control the ship's movement.
- Press the "W" key to accelerate the ship.
- Press the "S" key to brake the ship.
- Press the "A" and "D" keys to rotate the ship left and right.
- Press the "Esc" key to open the settings menu.
- Press the "R" key to respawn.

## Credits
- Game developed by Mirko Benincasa
- Background music: Battle Arcade by Panda Beats - No Copyright Music

## License
This project is licensed under the [GPL 3.0 License](https://www.gnu.org/licenses/gpl-3.0.html).

## Feedback and Issues
Have fun playing Starship Showdown! If you encounter any issues or have suggestions, feel free to [open an issue](https://github.com/MBenincasa/starship-showdown/issues).


CHANGELOG.md
# Changelog

## Version 0.2.0

### Features
- Added an AI-controlled Ship Rival that competes against the player, following checkpoints and providing challenging races.
- Introduced ship respawn functionality with a time penalty.

### Improvements
- Improved window size creation by dynamically adapting to monitor information, enhancing the gaming experience.



## Version 0.1.0

### Features
- Added basic spaceship controls for navigation.
- Implemented checkpoints for completing laps.
- Added lap counting and timer functionality.
- Included a debug mode for ship information display.
- Created a settings menu for toggling music, restarting the game, and exiting.
- Released both the .love file and standalone .exe for easy access.
- Added a README with instructions on how to play and download the game.

This marks the initial release of Starship Showdown, offering players a simple yet enjoyable racing experience.

main.lua
-- Imports
love.audio = require("love.audio")
local Game = require("src.game")
local Settings = require("src.settings")

-- Variables
local game
local backgroundMusic

-- This method is called only once at the beginning of the game's execution.
function love.load()
    -- love._openConsole()
    love.window.setTitle("Starship Showdown")

    -- Get the width and height of the current desktop resolution
    local desktopWidth, desktopHeight = love.window.getDesktopDimensions()

    -- Set the window mode to match the desktop resolution
    love.window.setMode(desktopWidth, desktopHeight, {
        fullscreen = true,  -- Set fullscreen mode
        resizable = true,   -- Allow window resizing
        vsync = true        -- Enable vertical synchronization
    })

    -- Set a larger font size
    local fontSize = 14
    local font = love.graphics.newFont(fontSize)
    love.graphics.setFont(font)

    -- Loads the audio file
    backgroundMusic = love.audio.newSource("resources/background.mp3", "stream")
    backgroundMusic:setLooping(true)
    -- Start audio playback
    love.audio.play(backgroundMusic)

    game = Game.new()
end

-- This method is called during every game cycle. It receives the parameter dt (delta time), which represents the time interval elapsed between the previous frame and the current one.
function love.update(dt)
    if not Settings.isMenuOpen then
        game:update(dt) -- Update only if the settings menu is not open
    end
    backgroundMusic:setVolume(Settings.isMusicEnabled and 1 or 0)
end

-- This method is called immediately after love.update() within the game cycle. This is where objects, images, and any visual elements on the screen are drawn.
function love.draw()
    game:draw()
end

-- This function is called whenever a key is pressed.
function love.keypressed(key)
    Settings:keypressed(key) -- Delegate the key press handling to the Settings module.
end


src/game.lua
-- Imports
local Ship = require("src.objects.ship")
local Rival = require("src.objects.rival")
local Checkpoint = require("src.objects.checkpoint")
local Settings = require("src.settings")

local Game = {} -- Declaration of an empty table named "Game"
Game.__index = Game -- Set the index of the "Game" table to itself

-- Constructor for creating a new Game instance
function Game.new()
    local self = setmetatable({}, Game) -- Create a new instance of the Game class
    self.ship = Ship.new() -- Initialize the ship object using the Ship class
    self.checkpoints = { -- Initialize the list of checkpoints
        Checkpoint.new(200, 200, 1),
        Checkpoint.new(400, 740, 2),
        Checkpoint.new(740, 300, 3)
    }
    self.rival = Rival.new(self.checkpoints) -- Initialize the rival's ship object using the Rival class
    self.currentCheckpointIndex = 1 -- Set the current checkpoint index to 1
    self.lap = 1 -- Initialize lap counter to 1
    self.rivalLap = 1 -- Initialize rival's lap counter to 1
    self.timer = 0 -- Initialize timer to 0
    self.rivalTimer = 0 -- Initialize rival's timer to 0
    self.isRunning = true -- Set the game to be running initially
    self.isRivalRunning = true -- Set the rival's game to be running initially
    self.maxLaps = 3 -- Set the maximum number of laps for the race
    self.checkpointTimes = {} -- Initialize an empty table to store checkpoint times for the ship
    self.rivalCheckpointTimes = {} -- Initialize an empty table to store checkpoint times for the rival
    self.lapTimes = {} -- Initialize an empty table to store lap times for the ship
    self.rivalLapTimes = {} -- Initialize an empty table to store lap times for the rival
    self.isFinished = false -- Flag to indicate if the ship's race is finished
    self.isRivalFinished = false -- Flag to indicate if the rival's race is finished
    return self -- Return the newly created Game instance
end

-- Update method for the Game class
function Game:update(dt)
    if self.isRunning then
        self.ship:update(dt) -- Update the ship's state based on the delta time (dt)
        self:checkCollisions() -- Check for collisions with checkpoints
        self.timer = self.timer + dt -- Increment the timer by the delta time

        if self.ship.isRespawned then
            self.timer = self.timer + (50 * dt)
            self.ship.isRespawned = false
        end
    end

    if self.isRivalRunning then
        self.rival:update(dt) -- Update the rival's ship
        self:checkRivalCollisions() -- Check for rival's collisions with checkpoints
        self.rivalTimer = self.rivalTimer + dt -- Increment the rival's timer by the delta time
    end

    if self.isRivalFinished then
        self.isRivalRunning = false -- Stop the rival's ship from updating when it's finished
    end

    if self.isFinished then
        self.isRunning = false -- Stop updating the ship when the race is finished
    end
end

-- Draw method for the Game class
function Game:draw()
    -- Check if the settings menu is open
    if Settings.isMenuOpen then
        Settings:drawSettingsMenu() -- If open, draw the settings menu.
    else
        Checkpoint.draw(self.checkpoints, self.currentCheckpointIndex) -- Draw checkpoints with currentCheckpointIndex indicating the next checkpoint
        self.ship:draw() -- Draw the ship
        self.rival:draw() -- Draw the rival's ship

        love.graphics.setColor(1, 1, 1)

        local yOffset = 60 -- Vertical offset for lap time display
        local lineHeight = 20 -- Height of each line
        local xOffset = 10 -- Horizontal offset for lap time display

        -- Display lap times and total lap time for Ship in descending order
        for lapIndex = #self.lapTimes, 1, -1 do
            local lapTimes = self.lapTimes[lapIndex]
            local lapTimeText = "LAP " .. lapIndex .. " - "
            for checkpointIndex, time in ipairs(lapTimes) do
                lapTimeText = lapTimeText .. "T" .. checkpointIndex .. ": " .. string.format("%.3f", time) .. "s "
            end
            lapTimeText = lapTimeText .. "- TOTAL: " .. self:getTotalLapTime(lapTimes) .. "s"
            local lapTimeX = xOffset
            local lapTimeY = 10 + (lapIndex - 1) * lineHeight
            love.graphics.print(lapTimeText, lapTimeX, lapTimeY)
        end

        -- Display lap times and total lap time for Rival in descending order
        for lapIndex = #self.rivalLapTimes, 1, -1 do
            local lapTimes = self.rivalLapTimes[lapIndex]
            local lapTimeText = "RIVAL LAP " .. lapIndex .. " - "
            for checkpointIndex, time in ipairs(lapTimes) do
                lapTimeText = lapTimeText .. "T" .. checkpointIndex .. ": " .. string.format("%.3f", time) .. "s "
            end
            lapTimeText = lapTimeText .. "- TOTAL: " .. self:getTotalLapTime(lapTimes) .. "s"
            local lapTimeWidth = love.graphics.getFont():getWidth(lapTimeText)
            local rivalLapTimeX = love.graphics.getWidth() - lapTimeWidth - xOffset
            local rivalLapTimeY = 10 + (lapIndex - 1) * lineHeight
            love.graphics.print(lapTimeText, rivalLapTimeX, rivalLapTimeY)
        end

        -- Display "Race Finished" if the race is finished
        if self.isFinished then
            love.graphics.print("Race Finished", xOffset, love.graphics.getHeight() - yOffset)
        end

        -- Display lap number in the top-center
        if not self.isFinished then
            local lapNumberText = "Lap: " .. self.lap .. "/" .. self.maxLaps
            local lapNumberWidth = love.graphics.getFont():getWidth(lapNumberText)
            local lapNumberX = (love.graphics.getWidth() - lapNumberWidth) / 2
            love.graphics.print(lapNumberText, lapNumberX, 10)
        end
    end
end

-- Method to check collisions with checkpoints for the Ship
function Game:checkCollisions()
    if self.isFinished then
        return
    end

    local currentCheckpoint = self.checkpoints[self.currentCheckpointIndex] -- Get the current checkpoint
    local distanceToCheckpoint = math.sqrt((self.ship.x - currentCheckpoint.x)^2 + (self.ship.y - currentCheckpoint.y)^2) -- Calculate the distance to the current checkpoint

    -- Check if the ship is close to the checkpoint and hasn't passed it already
    if distanceToCheckpoint <= currentCheckpoint.radius and not currentCheckpoint.passed then
        currentCheckpoint.passed = true -- Mark the checkpoint as passed
        table.insert(self.checkpointTimes, self.timer) -- Store the time taken to reach the checkpoint

        self.timer = 0 -- Reset the lap timer
        self.currentCheckpointIndex = self.currentCheckpointIndex + 1 -- Move to the next checkpoint

        if self.currentCheckpointIndex > #self.checkpoints then
            self.currentCheckpointIndex = 1 -- Wrap around to the first checkpoint if all checkpoints are passed
            self:completePlayerLap() -- Complete a lap when all checkpoints are passed
        end
    end
end

-- Method to check collisions with checkpoints for the Rival
function Game:checkRivalCollisions()
    if self.isRivalFinished then
        return
    end

    local rivalCurrentCheckpoint = self.rival.checkpoints[self.rival.currentCheckpointIndex] -- Get the current checkpoint for the Rival
    -- Calculate distances to the checkpoints for both Ship and Rival
    local distanceToCheckpoint = math.sqrt((self.rival.ship.x - rivalCurrentCheckpoint.x)^2 + (self.rival.ship.y - rivalCurrentCheckpoint.y)^2)
    -- Check if the rival is close to the checkpoint and hasn't passed it already
    if distanceToCheckpoint <= rivalCurrentCheckpoint.radius then
        table.insert(self.rivalCheckpointTimes, self.rivalTimer) -- Store the time taken by Ship to reach the checkpoint

        self.rivalTimer = 0 -- Reset the lap timer
        self.rival.currentCheckpointIndex = self.rival.currentCheckpointIndex + 1

        if self.rival.currentCheckpointIndex > #self.rival.checkpoints then
            self.rival.currentCheckpointIndex = 1
            self:completeRivalLap()
        end
    end
end

-- Method to complete a lap for the Ship
function Game:completePlayerLap()
    table.insert(self.lapTimes, self.checkpointTimes) -- Store the checkpoint times for the completed lap
    self.checkpointTimes = {} -- Reset checkpoint times for the next lap

    self.lap = self.lap + 1 -- Increment lap count
    self.timer = 0 -- Reset the lap timer
    self:resetCheckpoints() -- Reset checkpoint statuses for the new lap

    if self.lap > self.maxLaps then
        self.isFinished = true -- Finish the race when the maximum laps are completed
    end
end

-- Method to complete a lap for the Rival
function Game:completeRivalLap()
    table.insert(self.rivalLapTimes, self.rivalCheckpointTimes) -- Store the checkpoint times for the completed lap
    self.rivalCheckpointTimes = {} -- Reset checkpoint times for the next lap
    self.rivalLap = self.rivalLap + 1
    self.rivalTimer = 0 -- Reset the lap timer

    if self.rivalLap > self.maxLaps then
        self.isRivalFinished = true
    end
end

-- Method to reset checkpoint statuses for a new lap
function Game:resetCheckpoints()
    for _, checkpoint in ipairs(self.checkpoints) do
        checkpoint.passed = false -- Reset the "passed" status for all checkpoints
    end
end

-- Method to calculate the total lap time
function Game:getTotalLapTime(lapTimes)
    local totalLapTime = 0
    for _, time in ipairs(lapTimes) do
        totalLapTime = totalLapTime + time
    end
    return string.format("%.3f", totalLapTime) -- Return the total lap time formatted to three decimal places
end

return Game -- Return the Game table to make the class available for import and use in other files.

src/settings.lua
-- Create a table named "Settings" to hold settings-related functions and values.
local Settings = {}

-- Initialize settings variables.
Settings.isMenuOpen = false -- Flag to track if the settings menu is open.
Settings.isMusicEnabled = true -- Flag to track if background music is enabled.

-- Define a function to draw the settings menu on the screen.
function Settings:drawSettingsMenu()
    love.graphics.setColor(1, 1, 1) -- Set text color to white.

    local windowWidth, windowHeight = love.graphics.getWidth(), love.graphics.getHeight()
    local menuTitle = "Settings Menu"
    local optionToggleMusic = "1. Toggle Music: " .. (Settings.isMusicEnabled and "On" or "Off")
    local optionRestartGame = "2. Restart Game"
    local optionExitGame = "3. Exit Game"

    -- Calculate text dimensions
    local fontSize = 24
    local font = love.graphics.newFont(fontSize)
    love.graphics.setFont(font)

    local textHeight = love.graphics.getFont():getHeight()
    local titleWidth = love.graphics.getFont():getWidth(menuTitle)
    local optionWidth = math.max(
        love.graphics.getFont():getWidth(optionToggleMusic),
        love.graphics.getFont():getWidth(optionRestartGame),
        love.graphics.getFont():getWidth(optionExitGame)
    )

    -- Calculate positions for centering
    local titleX = (windowWidth - titleWidth) / 2
    local startY = (windowHeight - textHeight * 3) / 2 -- Center vertically

    love.graphics.print(menuTitle, titleX, startY) -- Draw the title of the settings menu.
    love.graphics.print(optionToggleMusic, (windowWidth - optionWidth) / 2, startY + textHeight) -- Display the option to toggle music on/off.
    love.graphics.print(optionRestartGame, (windowWidth - optionWidth) / 2, startY + textHeight * 2) -- Display the option to restart the game.
    love.graphics.print(optionExitGame, (windowWidth - optionWidth) / 2, startY + textHeight * 3) -- Display the option to exit the game.

    fontSize = 14
    font = love.graphics.newFont(fontSize)
    love.graphics.setFont(font)
end

-- Define a function to handle key presses for settings.
function Settings:keypressed(key)
    if key == "escape" then
        Settings.isMenuOpen = not Settings.isMenuOpen -- Toggle the settings menu open/close when "Esc" is pressed.
    elseif Settings.isMenuOpen then
        -- If the settings menu is open:
        if key == "1" then
            Settings.isMusicEnabled = not Settings.isMusicEnabled -- Toggle music on/off when "1" is pressed.
        elseif key == "2" then
            love.audio.stop() -- Stop the background music when music is disabled.
            love.load() -- Restart the game when "2" is pressed.
            Settings.isMenuOpen = false
        elseif key == "3" then
            love.event.quit() -- Exit the game when "3" is pressed.
        end
    end
end

-- Return the Settings table to make it accessible for use in other files.
return Settings


src/objects/rival.lua
local Ship = require("src.objects.ship")

local Rival = {} -- Declare an empty table named "Rival"
Rival.__index = Rival -- Set the index of the "Rival" table to itself

-- Constructor for creating a new Rival instance
function Rival.new(checkpoints)
    local self = setmetatable({}, Rival)
    self.checkpoints = checkpoints -- List of checkpoints to follow
    self.currentCheckpointIndex = 1 -- Current checkpoint index
    self.ship = Ship.new() -- Create a new ship instance for the rival
    self.color = {1, 0, 0} -- Set the color of the rival's ship
    return self
end

-- Update method for the Rival class
function Rival:update(dt)
    local currentCheckpoint = self.checkpoints[self.currentCheckpointIndex] -- Get the current checkpoint
    local currentCheckpointIndex = self.currentCheckpointIndex
    local dx = currentCheckpoint.x - self.ship.x
    local dy = currentCheckpoint.y - self.ship.y
    local distanceToCheckpoint = math.sqrt(dx^2 + dy^2)
    local angleToCheckpoint = math.atan2(dy, dx) + math.rad(90) -- Calculate the angle to the current checkpoint
    local angleToDirection = math.atan2(self.ship.velocity.y, self.ship.velocity.x) + math.rad(90) -- Calculate the angle to the ship's velocity direction
    local anglePositionDifference = (angleToCheckpoint - self.ship.rotation) % (2 * math.pi)  -- Calculate the angle difference between the ship's current rotation and the checkpoint
    local angleDirectionDifference = (angleToCheckpoint - angleToDirection) % (2 * math.pi) -- Calculate the angle difference between the checkpoint and the ship's velocity direction
    local speed = self.ship:getSpeed()

    -- When the distance to the checkpoint is small and the ship's speed is high, adjust orientation
    if distanceToCheckpoint < 310 and distanceToCheckpoint >= 40 and speed > 90 then
        if (angleDirectionDifference > math.pi) then
            self.ship:rotateToLeft(dt)
        elseif (angleDirectionDifference < math.pi or angleDirectionDifference > 0) then
            self.ship:rotateToRight(dt)
        end
    -- When close to the checkpoint, move to the next checkpoint
    elseif distanceToCheckpoint < 43 then
        currentCheckpointIndex = currentCheckpointIndex + 1
        if currentCheckpointIndex > #self.checkpoints then
            currentCheckpointIndex = 1
        end
        currentCheckpoint = self.checkpoints[currentCheckpointIndex]
        dx = currentCheckpoint.x - self.ship.x
        dy = currentCheckpoint.y - self.ship.y
        distanceToCheckpoint = math.sqrt(dx^2 + dy^2)
        angleToCheckpoint = math.atan2(dy, dx) + math.rad(90)
        anglePositionDifference = (angleToCheckpoint - self.ship.rotation) % (2 * math.pi)
        if (anglePositionDifference > math.pi) then
            self.ship:rotateToLeft(dt)
        elseif (anglePositionDifference < math.pi or anglePositionDifference > 0) then
            self.ship:rotateToRight(dt)
        end
    else
        -- Adjust orientation based on the angle between current orientation and desired direction
        if (anglePositionDifference > math.pi) then
            self.ship:rotateToLeft(dt)
        elseif (anglePositionDifference < math.pi or anglePositionDifference > 0) then
            self.ship:rotateToRight(dt)
        end
    end

    -- Apply braking or acceleration based on distance and speed
    if ((distanceToCheckpoint < 170 and distanceToCheckpoint >= 70 and speed > 110) or (distanceToCheckpoint < 70 and speed > 60)) then
        self.ship:brake(dt)
    else
        self.ship:accelerate(dt)
    end

    self.ship:calculateNextPosition(dt)
end

-- Draw method for the Rival class
function Rival:draw()
    love.graphics.setColor(self.color) -- Set the drawing color to the rival's color
    self.ship:draw() -- Draw the rival's ship using the Ship's draw method
    love.graphics.setColor(1, 1, 1) -- Reset the drawing color to white
end

return Rival -- Return the Rival table to make the class available for import and use in other files.

src/objects/checkpoint.lua
local Checkpoint = {} -- Declaration of an empty table named "Checkpoint"
Checkpoint.__index = Checkpoint -- Set the index of the "Checkpoint" table to itself

function Checkpoint.new(x, y, order)
    local self = setmetatable({}, Checkpoint)
    self.x = x
    self.y = y
    self.order = order -- Order/index of the checkpoint
    self.radius = 30 -- Radius of the checkpoint circle
    self.passed = false -- Flag indicating if the ship has passed this checkpoint
    return self
end

function Checkpoint.draw(checkpoints, currentCheckpointIndex)
    for _, checkpoint in ipairs(checkpoints) do
        local text = tostring(checkpoint.order)
        local textWidth = love.graphics.getFont():getWidth(text)
        local textHeight = love.graphics.getFont():getHeight()

        if checkpoint.order == currentCheckpointIndex then
            -- Draw the current checkpoint as a filled circle with a different color
            love.graphics.setColor(0.5, 0.5, 1)
            love.graphics.circle("fill", checkpoint.x, checkpoint.y, checkpoint.radius)

            -- Draw the text label at the center of the filled circle
            love.graphics.setColor(1, 1, 1)
            love.graphics.print(text, checkpoint.x - textWidth / 2, checkpoint.y - textHeight / 2)
        else
            -- Draw other checkpoints as outlined circles with the text label
            love.graphics.setColor(1, 1, 1)
            love.graphics.circle("line", checkpoint.x, checkpoint.y, checkpoint.radius)
            love.graphics.print(text, checkpoint.x - textWidth / 2, checkpoint.y - textHeight / 2)
        end
    end

    love.graphics.setColor(1, 1, 1) -- Reset the color
end

return Checkpoint


src/objects/ship.lua
local Ship = {} -- Declaration of an empty table named "Ship"
Ship.__index = Ship -- Set the index of the "Ship" table to itself

function Ship.new(x, y)
    local self = setmetatable({}, Ship) -- Create a new instance (object) of the Ship class
    self.x = x or 640
    self.y = y or 600
    self.velocity = { x = 0, y = 0 } -- Velocity vector
    self.acceleration = 160 -- Ship acceleration
    self.maxSpeed = 240 -- Maximum speed of the ship
    self.rotation = 0 -- Rotation angle of the ship in radians
    self.frictionCoefficient = 0.5 -- Friction coefficient
    self.turnSpeed = 90 -- Ship's turning speed (in radians per second)
    self.isRespawned = false
    return self
end

-- This update method defines the update behavior of the ship
function Ship:update(dt)
    if love.keyboard.isDown("left", "a") then
        self:rotateToLeft(dt)
    elseif love.keyboard.isDown("right", "d") then
        self:rotateToRight(dt)
    end

    if love.keyboard.isDown("up", "w") then
        self:accelerate(dt)
    elseif love.keyboard.isDown("down", "s") then
        self:brake(dt)
    end

    if love.keyboard.isDown("r") then
        self.x = 640
        self.y = 600
        self.rotation = 0
        self.velocity = { x = 0, y = 0 }
        self.isRespawned = true
    end

    self:calculateNextPosition(dt)
end

-- This draw method defines how the ship is drawn on the screen
function Ship:draw()
    love.graphics.push() -- This function saves the current graphics transformation matrix. It's useful when you want to apply temporary transformations (like rotation or translation) only to a specific part of the drawing, without affecting the rest of the scene
    love.graphics.translate(self.x, self.y) -- This function translates the origin of the graphics coordinates so that it corresponds to the (self.x, self.y) position of the ship. In other words, this instruction moves the coordinate system so that the center of the ship is at coordinates (0, 0)
    love.graphics.rotate(self.rotation) -- This function rotates the coordinate system counterclockwise by an angle of self.rotation in radians
    love.graphics.polygon("fill", 0, -25, -15, 25, 15, 25) -- This instruction draws a filled polygon that represents the ship
    love.graphics.pop() -- This function restores the previous graphics transformation matrix that was saved with push(). This is important to ensure that the transformations don't affect other objects drawn after the ship
end

function Ship:debug()
    local speed = self:getSpeed()
    local directionAngle = self:getDirectionAngle()
    local debugText = {
        "Debug Info:",
        "x: " .. string.format("%.2f", self.x),
        "y: " .. string.format("%.2f", self.y),
        "acceleration: " .. string.format("%.2f", love.keyboard.isDown("w", "up") and self.acceleration or 0) .. " px/s²",
        "brake: " .. (love.keyboard.isDown("s", "down") and "Pressed" or "Not pressed"),
        "rotation: " .. string.format("%.2f", self.rotation % 6.28) .. " rad",
        "velocity: x=" .. string.format("%.2f", self.velocity.x) .. ", y=" .. string.format("%.2f", self.velocity.y),
        "speed: " .. string.format("%.2f", speed) .. " px/s",
        "directionAngle: " .. string.format("%.2f", directionAngle) .. "°",
        "frictionCoefficient: " .. self.frictionCoefficient
    }

    local lineHeight = 20
    local windowWidth, windowHeight = love.graphics.getWidth(), love.graphics.getHeight()

    love.graphics.setColor(1, 1, 1)
    for i, textLine in ipairs(debugText) do
        local x = windowWidth - love.graphics.getFont():getWidth(textLine) - 10
        local y = windowHeight - 10 - (#debugText - i + 1) * lineHeight
        love.graphics.print(textLine, x, y)
    end
end

-- Calculate and return the magnitude of the velocity vector, representing the speed of the ship
function Ship:getSpeed()
    return math.sqrt(self.velocity.x^2 + self.velocity.y^2)
end

-- Calculate and return the direction of the ship in radians based on its velocity vector
function Ship:getShipDirection()
    return math.atan2(self.velocity.y, self.velocity.x);
end

-- Calculate and return the angle of direction (orientation) of the ship in degrees, considering its current rotation
function Ship:getDirectionAngle()
    -- Calculate the ship's direction angle based on its velocity direction and rotation
    local shipDirection = self:getShipDirection()
    local directionAngle = (math.deg(shipDirection - self.rotation) + 90) % 360
    -- Adjust the angle to be within the range [-180, 180]
    if directionAngle > 180 then
        directionAngle = directionAngle - 360
    end

    return directionAngle
end

function Ship:accelerate(dt)
    -- Calculate acceleration multiplied by delta time (dt)
    local acceleration = self.acceleration * dt
    -- Calculate acceleration along the x and y axes using trigonometric functions
    local accelerationX = math.sin(self.rotation) * acceleration
    local accelerationY = -math.cos(self.rotation) * acceleration
    -- Updates x and y velocity of ship with calculated acceleration
    self.velocity.x = self.velocity.x + accelerationX
    self.velocity.y = self.velocity.y + accelerationY
end

function Ship:brake(dt)
    local deceleration = self.acceleration * 0.005 * dt -- Calculate deceleration based on ship's acceleration and delta time
    local dotProduct = self.velocity.x * math.sin(self.rotation) + self.velocity.y * -math.cos(self.rotation) -- Calculate dot product of velocity and forward direction
    local directionAngle = self:getDirectionAngle()

    -- Calculate the factor by which to scale the deceleration based on direction angle
    local decelerationFactor = 1 - math.abs(directionAngle) / 90

    if dotProduct > 0 then
        -- Apply deceleration in the forward direction
        self.velocity.x = self.velocity.x - self.velocity.x * deceleration * decelerationFactor
        self.velocity.y = self.velocity.y - self.velocity.y * deceleration * decelerationFactor
    else
        -- Apply deceleration in the backward direction
        self.velocity.x = self.velocity.x + self.velocity.x * deceleration * decelerationFactor
        self.velocity.y = self.velocity.y + self.velocity.y * deceleration * decelerationFactor
    end
end

function Ship:rotateToRight(dt)
    self.rotation = self.rotation + math.rad(self.turnSpeed) * dt -- Rotates the ship to the right by adding a certain angle in radians
end

function Ship:rotateToLeft(dt)
    self.rotation = self.rotation - math.rad(self.turnSpeed) * dt -- Rotates the ship to the left by subtracting a certain angle in radians
end

function Ship:calculateNextPosition(dt)
    -- Calculate friction based on current velocity (always present)
    local frictionX = -self.velocity.x * self.frictionCoefficient
    local frictionY = -self.velocity.y * self.frictionCoefficient

    -- Add friction to velocity
    self.velocity.x = self.velocity.x + frictionX * dt
    self.velocity.y = self.velocity.y + frictionY * dt

    local minSpeed = 0 -- Limits the minimum speed due to friction
    -- Calculate the total speed (modulo of the velocity vector)
    local speed = self:getSpeed()
    if speed < minSpeed then
        self.velocity.x = 0
        self.velocity.y = 0
    end

    -- If the speed is higher than the maximum speed allowed, it reduces the speed to a fraction of the maximum speed
    if speed > self.maxSpeed then
        local scaleFactor = self.maxSpeed / speed
        self.velocity.x = self.velocity.x * scaleFactor
        self.velocity.y = self.velocity.y * scaleFactor
    end

    -- Update ship's x and y position based on current velocity
    self.x = self.x + self.velocity.x * dt
    self.y = self.y + self.velocity.y * dt
end

-- This line returns the Ship table to make the class available for import and use in other files.
return Ship